### ISSUE TITLE
Fully generalizing the rendering system

### ISSUE THREAD

**Issue by jlstevens** [ioam/holoviews#141]
While discussing the upcoming SciPy paper, Philipp and I realized that we have a few more steps to implement before we can claim that the ``HoloViews`` rendering system is entirely independent of matplotlib.

The main problem is that in the display hooks, we still pass around matplotlib figure and animation objects which are a matplotlib specific concept. We have already introduced the concept of a renderer to try and abstract the plotting backend from the rest of HoloViews. This process of abstraction is not entirely complete, and currently only one renderer may be set on the ``Store`` at a time.

Here is the new proposal:

* The current baseclass called ``Plot`` should be called ``MPLPlot``. This will allow ``Plot`` to be a general base class suitable for any rendering system, not just matplotlib.
* Get rid of ``Store.renderer`` and let each plot class have a ``.renderer`` attribute. This would allow you do use multiple renderers to render different elements in the same notebook! E.g one plot could use Cairo and the next one could use matplotlib.
* Due to import issues relative to ``__init__``, each backend needs a file called ``renderer.py`` that contains the renderer (to be accessible from the Plotting classes).

These changes mean that ``Store`` still holds the association from element type to plotting class in the registry, but you don't know what rendering system is associated with each Plot. 

The next issue is that ``__call__`` and ``anim`` methods on the current ``Plot`` classes return matplotlib figure and animation objects. These are meaningless concepts outside of matplotlib which means a different return type must be used.  We propose that ``self`` is returned where the purpose of the call is to update the state of the plotting class.

The obvious next question is how to generate output if the plot classes no longer return matplotlib figure/animation objects. For this we propose that plot classes have a ``.state`` (or ``.data``) attribute representing the state required by the renderer to generate meaningful output. For instance, in a Cairo renderer, this would probably be a Cairo surface. 

We would also need one new method to be supported by all renderers:

```python
renderer.render(plot_instance, key=None, fig_format=None, holomap_format=None)
```

This is the method that takes a plot instance to generate output from it. If key is supplied, ``__getitem__`` is called on the plot instance to update its state accordingly. If the formats are not supplied, they are taken from the class/instance parameters (renderers are ParameterizedFunctions). Note that because a plot instance is supplied, a single plot instance can be used to generate many frames without needing to create a new plot instance internally per frame.

The interface for the new ``Plot``, renderer-agnostic base class is as follows:

* ``__init__``:   Takes an element and plot parameters.
* ``state``:  The plotting state that has been drawn by the plot, ready to be turned into some useful output by the renderer.
*  ``__call__``:  Renders the element returning self with ``.state`` set such that the renderer can generate suitable output. When multiple frames are involved, this is always the last frame.
* ``__getitem__``: Select a key (i.e frame) from the supplied element and update  ``.state`` 
* ``__len__``: The number of frames available given the element.

This removes all mention of matplotlib-specific objects from the design. Note that ``MPLPlotRenderer`` will know about the ``MPLPlot.anim`` method in order to render gif, mp4 and other video formats. This is matplotlib specific as the notion of an animation object is not necessarily a general one: the proposed API for ``Plot`` is sufficient to generate animation by stitching a series of requested frames together.

Although this sounds like a major refactor, I don't think it will be so bad and I think it can be broken down into stages:

1. Split ``Plot`` into ``Plot`` and ``MPLPlot``.
2. Create ``renderer.py``, declare the renderer attribute on the plots and make use of it, removing ``Store.renderer``.
3. Implement the ``.state`` attribute and get ``MPLPlotRenderer`` to use it (no need to change the return types from the plots yet)
4. Switch the display hooks over to only use the renderer methods (including the new ``renderer.render`` method).
5. Return ``self`` from the ``__call__`` and ``anim`` methods.

I think we can have a working version of HoloViews that doesn't break the tests at each of these stages. It will require a bit of work but something we want to fix before 1.2.1 and the SciPy paper!

**Comment by philippjfr** [comment]
I think this a good summary of our discussion. The steps also seem to be correct and in the right order, although I'm sure something else will crop up when we actually try to implement it.

**Comment by jlstevens** [comment]
Some wrinkles to worry about in the display hooks:

* There is some logic setting axes for 3D elements (maybe the plots themselves can do this?)
* Apparently mpld3 needs some special handling (``fig.dpi`` is used for instance)

Otherwise, I still think the plan should be fine.

**Comment by jbednar** [comment]
That all sounds good; it seems very important to do, and not too difficult.

**Comment by jlstevens** [comment]
This was a fair bit of effort to do but it is mostly done now. 

There is still more tidying up required and I also need to actually think about the status of the ``animation_data`` method on ``MPLRenderer`` as it is still not yet part of the generalized ``Renderer`` API. In theory, we could now have every other plot rendered with cairo instead of matplotlib!

Otherwise, the display hooks are finally free from matplotlib except for things involving the 'd3' or 'nbagg' backends. I am no longer sure how I feel about these 'backends' set via the output magic - it seems to me that this will break the idea that each ``Plot`` class can have a different renderer (that do not involve matplotlib!)

One thing I am wondering is whether the set of formats suggest by the output magic should now be fixed: without a plot class you have no idea what renderer will be used and therefore no idea which formats may or may not be supported. I think it might make sense just to have a list of formats that make sense in the browser and it is up to the individual renderers to either support what is requested or throw an exception.

As of 395ca51 all these changes seem to be passing on Travis and the notebooks I've tested seem fine (though I think Philipp mentioned that nbagg and mpld3 output is no longer centered for some reason).

**Comment by jlstevens** [comment]
The new system mostly seems to be working but there is still the issue of selecting the backends (issue #142 ) and the problem that the ``Store.registry`` is currently only populated with matplotlib plots with nowhere to register plot classes from other backends.

I think we should register *all* the available plot classes regardless of backend in the ``Store`` registry. Using bokeh as a possible new backend along with matplotlib, you can imagine something like this:

```python
Store.registry = {'maplotlib':dict(Layout=plotting.mpl.LayoutPlot,
                                   Curve=plotting.mpl.CurvetPlot),
                  'bokeh':dict(Layout=plotting.bokeh.LayoutPlot,
                               Curve=plotting.bokeh.CurvetPlot) }
```

This would help alleviate the problems of composition because if you have two bokeh ``CurvePlot`` instances (say) you will only be able to compose them using a bokeh ``LayoutPlot`` and not the matplotlib equivalent. Storing all the available plot classes means each backend could then have access to all the other plot classes belonging to the same backend which would avoid complicating the implementation of the plot classes such as ``OverlayPlot`` or ``LayoutPlot``.

The last thing to do is then to address issue #142 : I now think we should use the backend argument of the output magic to select the actual backend (e.g 'matplotlib', 'bokeh', 'cairo') instead of the current options ('mpl', 'd3' or 'nbagg'). These latter options are renderer sub-options which could be addressed with a new renderer magic or perhaps more simply by setting the backend value to a string such as 'matplotlib:nbagg' or 'matplotlib:d3'.

In the end, I think I am not in favor of introducing a new magic:  I would prefer re-purposing the 'backend' argument of the output magic as suggested above. Of course, this assumes that in an interactive session, the user may want to select from a small set of possible renderer 'modes'. An ``ObjectSelector`` parameter called 'modes' on the renderer classes could make this approach work fairly easily.

Naturally, this approach would not scale to advanced renderer customization but they should probably be achieved by setting renderer parameters appropriately at the class level anyway.

Thoughts?

**Comment by philippjfr** [comment]
That all sounds reasonable. I've already run into this issue while implementing the bokeh plots. I don't think another magic is necessary either, re-purposing the output magic makes sense. We need to consider that different backends only support specific figure formats and set them dynamically. We also need the OptionTrees to be switched out dynamically so you can switch back and forth between backends and have the correct options become available as appropriate.

**Comment by jlstevens** [comment]
Right, I will come back to the ``OptionTree`` issue later...

What I want to suggest now are the following improvements in addition to my suggestions above (the registry containing all the plot classes):

* ``Store.renderer`` should be resurrected as  ``Store.renderers`` - a dictionary of backend names to the corresponding renderer class  (e.g ``Store.renderers['matplotlib']``). This is so the user can get the renderer easily and the output magic could inspect the renderers for their names and supporter modes (e.g to allow auto-completion of the proposed 'backend' option in the output magic).

* I no longer think it is a good idea to pull the concrete classes into ``holoviews.plotting``. Instead, I think we should have a ``plotclass`` method the renderer base class that when given an element returns the appropriate plotting class. This makes sense as we already require a ``plot_options`` method. 

Together with the suggestion above, you can do ``Store.renderers['matplotlib'].plotclass(image)`` to get the matplotlib version of the appropriate plotting class (i.e ``RasterPlot``). This system doesn't stop you using imports directly (e.g ``from plotting.mpl import RasterPlot``) and we could add a helper method to ``Store`` if we *really* want easy access to the plotting classes (I not in favor of introducing new methods unless we really need them).

I think these are good ideas which I am going to start implementing now - I believe these are separate issues from what needs to happen with ``OptionTrees`` (though I wonder if the trees could live on the renderer classes?)

**Comment by jlstevens** [comment]
Regarding the option trees:

* In the end, I think we should keep them on ``Store`` and simply qualify the backend both for the default trees and the custom trees. **Note:** This affects the pickling of styles!
* Some important methods on ``Store`` will need the backend as an argument e.g ``lookup_options``. We could make the backend a keyword argument (default 'matplotlib') for backwards compatibility. Alternatively, the cleaner interface is something like ``lookup_options('matplotlib', image, 'style').
* If we go with the latter, we can keep the plotting code clean by defining something like this on the ``Plot`` base class:

```python
def lookup_options(self, obj, group):
    return Store.lookup_options(self.renderer.backend, obj, group)
```
* Customizations done via functions (i.e ``custom_options(options): ...``) can remain backward compatible as long as the renderer is specified when registering them with the ``Store``.

**Comment by philippjfr** [comment]
>I am not happy with the way customizations are supplied in functions that are then registered with the Store anyway. That aspect probably needs a rethink and we can probably come up with a better approach anyway. Better think about that before trying to make the existing approach backwards compatible!

There are multiple ways of setting the OptionTree, I think this demonstrates all of them:

```python
options.Path = Options('style', linewidth=0.5)
options.Path = (Options('style', linewidth=0.5), Options('plot', xaxis=None))
options.Path = {'style': Options(linewidth=0.5), 'plot': dict(xaxis=None)}
```

I'd say this one:

```python
options.Path = (Options('style', linewidth=0.5), Options('plot', xaxis=None))
```

is particularly weird, which makes the situation awkward.

```python
options.Path = Options('style', linewidth=0.5)
```

is nicer than

```python
options.Path = {'style': dict(linewidth=0.5)}
```

but 

```python
options.Path = {'style': dict(linewidth=0.5), 'plot': dict(xaxis=None)}
```

is nicer than:

```python
options.Path = (Options('style', linewidth=0.5), Options('plot', xaxis=None))
```

Not sure what to suggest.

**Comment by jlstevens** [comment]
That is a separate issue that we can also think about later. I was talking about this mechanism of setting a callback:

```python
def default_options(options):
    # Charts
    options.Curve = Options('style', color=Cycle(), linewidth=2)
    ...


# Register the default options
Store.option_setters.append(default_options)
```

**Comment by jlstevens** [comment]
Ok, we've discussed this and I think we've identified the problem with our current approach. Currently, we have two steps:

* Add an ``element:plotclass`` to the registry dictionary directly.
* Call ``Store.register_plots`` which rebuilds the option tree from scratch each time it is called.

Because of the whole tree is rebuilt when ``register_plots`` is called (in order to get the allowed plot/style options), we then need these annoying callbacks to re-add the desired customizations.

The trick we need is to update the options tree incrementally instead of erasing it and rebuilding it. This should happen when an element and plot classes are added to the registry. In other words, we need a method that works like this:

```python
Store.register({Image:RasterPlot}, backend='matplotlib')
```

Not only does this qualify by the backend, this can incrementally update the option tree to make sure that only the appropriate plot and style options of ``RasterPlot`` are allowed.

I think this resolves these issues though there is one caveat: you must register your plots with the ``Store`` before trying to customize options.

**Comment by philippjfr** [comment]
This is definitely better and it means we can get rid of the option_setter callbacks. Hopefully not too many people have started using them, they haven't really been around for very long afterall.

I think for setting the options we've decided we'll just have separate trees for each backend, so to set option defaults you'll have to set it like this:

```python
Store.options['bokeh'].Image = Options('style', palette='Spectral9')
Store.options['matplotlib'].Image = Options('style', cmap='hot')
```

**Comment by jlstevens** [comment]
Sure, though I would recommend this style throughout:

```python
options = Store.options['bokeh']
options.Image = Options('style', palette='Spectral9')
```

and then:

```python
options = Store.options['matplotlib']
options.Image = Options('style', cmap='hot')
```

Should help avoids pointless repetition...

**Comment by jlstevens** [comment]
The proposed changes (not involving the options system) have now been implemented since commit 2eabeb9.

* ``Store.renderers`` now exists with only 'matplotlib' supported right now.
* ``Store.registry`` can now hold plot classes across multiple backends.
* ``Store.register`` is now the way of registering new plots with 'matplotlib' as the default backend (others can be specified).
* ``Renderer.plotting_class`` takes an element and returns the appropriate plotting class for it.
* There are no longer any concrete plotting classes in ``plotting`` itself. Instead, ``__init__.py`` simply imports all the available renderers to allow each backend to register itself with the ``Store`` in the appropriate way.

This isn't finished yet: I've added a temporary ``BACKEND`` global variable to the display hooks that will later be set by the output magic. In addition, ``holoviews.help`` now has a backend argument (default with a default of 'matplotlib') in order to select the appropriate plotting class for a given backend.

**Comment by jlstevens** [comment]
In commit 67ab735, I've got the display hooks and the output magic connected to enabled backend switching. A few things are still not finished.

* The 'nbagg' and 'd3' logic needs to be implemented on ``MPLRenderer`` instead of in the display hooks and widgets code. I hope @philippjfr will be able to help with this - until then these options won't work.

* I removed the following logic from the output magic for the time being:


```python
if options['backend'] == 'd3':
    try:      import mpld3 # pyflakes:ignore (Testing optional import)
    except:
        raise ValueError("Cannot use d3 backend without mpld3. "
                         "Please select a different backend")
    allowed = ['scrubber', 'widgets', 'auto']
    if options['holomap'] not in allowed:
        raise ValueError("The D3 backend only supports holomap options %r" % allowed)
```

**Comment by jlstevens** [comment]
With the output magic and widgets done, I think there are only two things left to do:

* Update the options system accordingly (shouldn't be too bad)
* Figure out how to make ``GridPlot`` and ``RasterGridPlot`` switch properly

**Comment by jlstevens** [comment]
The options system update is complete leaving the ``GridPlot`` and ``RasterGridPlot`` issue. We also want to update the ``Store.register_plots`` method to avoid the need of option setting callbacks.

**Comment by philippjfr** [comment]
I've just resolved the ``GridPlot``/``RasterPlot`` with a ``PlotWrapper`` class letting you define a function that will pick the plot based on the plotted object and exposes the plot and style options to the ``Store`` object. That just leaves the issues with ``register_plots``.

**Comment by philippjfr** [comment]
I think we've now resolved all the issues. The bokeh backend I've been working on seems to hook into the system seamlessly, which should be sufficient to demonstrate that the system is now backend independent. After the v1.2.5 release I'll add the new backend so we can improve it gradually. I'll close this issue now as we the design is in place and any further tweaks can become separate issues.

**Comment by jbednar** [comment]
That's great!  We're up to 1.2.5 now? :-)

**Comment by jlstevens** [comment]
Yes..very nearly.

We have a few minor things to clear up that should be very easy. I would also like to give us a few more days to find bugs after this (relatively large) refactor so I propose we release 1.2.5 on Monday.

**Comment by jbednar** [comment]
I'd vote for calling it 1.3, since it has so many changes relative to 1.2.0 by now.  That way 1.3.1 will be the one that actually works. :-)

**Comment by jlstevens** [comment]
The trend seems to be that we bump by 0.1 every release with 0.0.1 bumps for immediate fixes (i.e within a day) when things go wrong with the PyPI upload (normally for critical fixes).

Should we just make this our convention and give up on the idea of any other kind of release? I can imagine an x.x.2 release if we ever need two critical updates.

**Comment by jbednar** [comment]
Seems ok to me, though I don't think it matters whether the critical fixes are within a day.   It doesn't mean giving up on the idea of any other kind of release, because 2.0 can still be used for an even more major compatibility breaking change.

**Comment by jlstevens** [comment]
Well, once 1.3 is out, this would mean that we only have 6 more releases before hitting version 2.0. :-)

**Comment by jbednar** [comment]
Not really; you can have 1.12.0 if there's been no need for a 2.0 release by that time...

**Comment by jlstevens** [comment]
I guess that would work fine even if at a quick glance it is a little confusing (e.g could be confused with 1.1.2 but I guess that is fairly unlikely).
